File: tools/provisioning_lifecycle_cases.py
#!/usr/bin/env python3
"""Production provisioning lifecycle/capture cases for Border 179.

The installed Kahlua VM executes WorldSources, SourceUse, Needs, Controller and
decision_capture. Native holders and vanilla queue/actions are explicit adapters;
native inventory fidelity is covered by WorldTransferProbe.
"""
import json
import pathlib
import re
import shutil
import subprocess
import tempfile

ROOT = pathlib.Path(__file__).resolve().parents[1]
CASES = ROOT / "tools/sweep/provisioning_lifecycle_cases.lua"
CAPTURE = ROOT / "tools/sweep/decision_capture.lua"
PERCEPTION = ROOT / "mod/42.20/media/lua/shared/SAO_Perception.lua"
ATTACHMENT = ROOT / "mod/42.20/media/lua/shared/SAO_PlaceAttachment.lua"

EXPECTED = {
    "acquire_performed", "store_performed", "no_consumption_or_need_credit",
    "queued_is_not_performed", "queue_without_move_releases", "rejected_enqueue",
    "interrupt_before_move", "interrupt_after_move", "reload_before_move",
    "reload_after_move_once", "unavailable_destination_pending",
    "both_holders_conflict", "neither_holder_conflict", "standing_revocation_prevents_mutation",
    "acquired_contents_mismatch", "physical_anchor_refresh",
    "completed_work_credit_once", "credit_actor_status_order",
    "queue_has_no_social_credit", "capture_transfer_bound",
    "capture_decision_frozen", "capture_item_units",
    "capture_required_result_failure", "capture_required_person_failure",
    "standing_food_refusal", "desperate_food_admission", "legacy_medication_route",
    "legacy_missing_context",
    "nearby_approach_exact", "nearby_route_refusal", "nearby_standing_refusal",
    "nearby_reached_transfer",
    "production_inspection_geometry", "production_anchor_recall",
    "production_anchor_exact_scope", "production_consume_anchor",
    "production_inspection_no_attachment", "production_anchor_access_route",
    "production_building_visit_preserved",
}

ADAPTERS = r'''
plainNameOf = function(first, last) return first .. " " .. last end
SAO.History.ageOf = function() return 31 end
SAO.Disposition = {circle=function() return "near" end,
    traits=function() return {patience=0.7} end, describe=function() return "fixture" end}
SAO.Conditions = {of=function() return {tired=false} end}
SAO.Habits = {of=function() return {keepsWatch=true} end}
SAO.Lessons = {renderClaims=function() return {shelter=true} end}
SAO.Census = {JOB_PERK={forager="Foraging"}, classOf=function() return "farmer" end,
    skillOf=function() return 2 end}
SAO.Rand = {state=function() return "lifecycle-seed", 17 end}
SAO.Voice = {onEvent=function() __voiceCalls = __voiceCalls + 1 end}
SAO.Standing.adjustTrust = function() __trustCalls = __trustCalls + 1 end
SAO.Standing.addDebt = function() __debtCalls = __debtCalls + 1 end
SAO.Standing.settleDebt = function() __debtCalls = __debtCalls + 1 end
SAO.Standing.mayTakeCurrent = function(id,x,y,admission)
    __standingAdmission = admission
    return admission == "desperate" or __standingAllowed ~= false
end
SAO.Standing.mayAttemptBelieved = function(id,x,y,admission)
    __attempt = {id=id,x=x,y=y,admission=admission}
    return __attemptAllowed ~= false
end
SAO.Perception.beliefs = {}
SAO.Perception.learnInspectedSource = function(id, place, sourceId, tick, provenance)
    local actor = tostring(id)
    __known[actor] = __known[actor] or {}
    local belief = __known[actor][place.id] or {cx=place.cx,cy=place.cy,sourceFacts={}}
    __known[actor][place.id] = belief
    belief.sourceFacts[sourceId] = SAO.WorldSources.beliefFact(sourceId)
    belief.at, belief.source = tick, provenance
    __lastLearn = {place=place,sourceId=sourceId,provenance=provenance}
    return true
end
ISInventoryTransferAction = {}
function ISInventoryTransferAction:derive(name)
    local child = {}; child.__index = child
    setmetatable(child, {__index=self}); return child
end
function ISInventoryTransferAction:new(character,item,src,dest)
    return setmetatable({character=character,item=item,srcContainer=src,destContainer=dest},
        {__index=self})
end
function ISInventoryTransferAction:isValid() return __nativeValid ~= false end
function ISInventoryTransferAction:transferItem(item)
    __nativeMoves = __nativeMoves + 1
    __move()
end
ISTimedActionQueue.add = function(action)
    __queueCalls = __queueCalls + 1
    if __queueReject then return end
    __queued, __busy = action, true
end
ISTimedActionQueue.hasAction = function(action) return __queued == action and __busy end
local eatNew, drinkNew = ISEatFoodAction.new, ISDrinkFluidAction.new
ISEatFoodAction.new = function(...)
    __consumeCalls = __consumeCalls + 1; return eatNew(...)
end
ISDrinkFluidAction.new = function(...)
    __consumeCalls = __consumeCalls + 1; return drinkNew(...)
end
SAOJavaBridge.hasPendingActions = function() return __busy end
SAOJavaBridge.containerAccessibleNow = function() return __accessible end
SAOJavaBridge.worldTransferPosition = function() return "AT:8:8:0" end
SAOJavaBridge.worldTransferOffer = function() return __offerText end
SAOJavaBridge.carriedWorldTransferItem = function() return __carriedText end
SAOJavaBridge.worldStoreTransferState = function()
    if __holderState == "UNAVAILABLE" then return "UNAVAILABLE" end
    if __carriedItem and not __destinationItem then return "CARRIED" end
    if __destinationItem and not __carriedItem then return "TRANSFERRED" end
    return "CONFLICT"
end
SAOJavaBridge.findFoodSource = function(self,body,radius)
    __searchedRadius = radius; return __discoveredSource
end
SAOJavaBridge.foodSourceWithinReach = function() return __withinReach ~= false end
SAOJavaBridge.foodSourceItem = function() return __sourceItem end
SAOJavaBridge.foodSourceContainer = function() return __sourceContainer end
SAOJavaBridge.bindWorldStoreAction = function()
    __bindCalls = __bindCalls + 1; return __bindAnswer
end
SAOJavaBridge.bindWorldSourceAction = function()
    __bindCalls = __bindCalls + 1; return __bindAnswer
end
SAOJavaBridge.worldSourceActionOriginContainer = function() return __originContainer end
SAOJavaBridge.grantXP = function(self,body,perk,amount)
    __xpCalls = __xpCalls + 1; __xpPerk = perk; __xpAmount = amount
end
'''


def _dependencies():
    # Border 179 imports this module; defer its reciprocal fixture import.
    import source_use_test as source
    import provisioning_transfer_cases as transfers
    return source, transfers


def _prelude(source, transfers):
    values = dict(transfers.SNAPSHOTS)
    values.update(store_row=transfers.transfer_row(),
                  acquire_row=transfers.transfer_row("acquire"),
                  anchor_accessible=transfers.SNAPSHOTS["acquire_pre"].replace(
                      "access=unknown", "access=accessible", 1))
    encoded = "\n".join(f"__{key} = {json.dumps(value)}" for key, value in values.items())
    return source.ACTION_PRELUDE + "\n" + ADAPTERS + "\n" + encoded


def _run(work, source, transfers, overrides=None):
    overrides = overrides or {}
    prelude = work / "prelude.lua"
    prelude.write_text(_prelude(source, transfers), encoding="utf-8")
    chunks = [prelude]
    for path in (source.WORLD, source.NEEDS, source.SOURCE_USE,
                 source.CONTROLLER, CAPTURE):
        text = overrides.get(path.name)
        if text is not None:
            destination = work / path.name
            destination.write_text(text, encoding="utf-8")
            chunks.append(destination)
        else:
            chunks.append(path)
    probe = work / "cases.lua"
    probe.write_text("__lifecycleResult = " + CASES.read_text(encoding="utf-8"),
                     encoding="utf-8")
    chunks.append(probe)
    # Load the real producer after adapter-based lifecycle coverage. These
    # cases cannot accidentally exercise the earlier perception test double.
    for path in (PERCEPTION, ATTACHMENT):
        text = overrides.get(path.name)
        if text is not None:
            destination = work / path.name
            destination.write_text(text, encoding="utf-8")
            chunks.append(destination)
        else:
            chunks.append(path)
    anchor_probe = work / "perception-cases.lua"
    anchor_probe.write_text("__productionPerceptionCases = true\n"
        "__lifecycleResult = __lifecycleResult .. '|' .. "
        + CASES.read_text(encoding="utf-8"), encoding="utf-8")
    chunks.append(anchor_probe)
    done = subprocess.run([str(source.JDK / "java.exe"), "-cp", f"{source.PZ};.",
        "LuaRun", *map(str, chunks), "--", "__lifecycleResult"],
        cwd=work, capture_output=True, text=True, timeout=120)
    detail = (done.stdout or "") + (done.stderr or "")
    lines = (done.stdout or "").strip().splitlines()
    value = lines[-1][6:] if lines and lines[-1].startswith("VALUE ") else ""
    checks = dict(re.findall(r"([a-z0-9_]+)=(true|false)", value))
    return checks, detail


def _compile(work, source):
    shutil.copy2(source.STDLIB, work / "stdlib.lua")
    compiled = subprocess.run([str(source.JDK / "javac.exe"), "-cp", str(source.PZ),
        "-d", str(work), str(source.RUNNER)], capture_output=True, text=True, timeout=120)
    if compiled.returncode:
        raise RuntimeError((compiled.stderr or compiled.stdout)[-2000:])


def run_cases(overrides=None):
    source, transfers = _dependencies()
    with tempfile.TemporaryDirectory(prefix="sao-transfer-lifecycle-") as temporary:
        work = pathlib.Path(temporary)
        _compile(work, source)
        return _run(work, source, transfers, overrides)


MUTATIONS = (
    ("authority-at-transfer", "SAO_Needs.lua",
     "if not self:isValid() then self.dontAdd = true; return end", "if false then return end",
     "standing_revocation_prevents_mutation"),
    ("performed-holder-proof", "SAO_SourceUse.lua",
     'if state == "TRANSFERRED" then', 'if state == "TRANSFERRED" or state == "CARRIED" then',
     "queue_without_move_releases"),
    ("work-credit-idempotence", "SAO_Controller.lua",
     'if not record or order <= (tonumber(record.provisioningCreditOrder) or 0) then',
     'if not record then', "completed_work_credit_once"),
    ("transfer-capture-binding", "decision_capture.lua",
     'if reservation then event.reservationId = reservation.id',
     'if reservation then event.unboundReservationId = reservation.id',
     "capture_transfer_bound"),
    ("inspection-anchor-geometry", "SAO_Perception.lua",
     "belief.cx, belief.cy, belief.z = place.cx, place.cy, place.z",
     "belief.x, belief.y, belief.z = place.cx, place.cy, place.z",
     "production_inspection_geometry"),
    ("inspection-social-separation", "SAO_Perception.lua",
     "if not belief.sourceId then places[key] = belief end",
     "if true then places[key] = belief end",
     "production_inspection_no_attachment"),
)


def main():
    source, transfers = _dependencies()
    if not all(path.is_file() for path in (source.PZ, source.STDLIB,
            source.JDK / "java.exe", source.JDK / "javac.exe")):
        print("Provisioning lifecycle SKIPPED: installed game VM or JDK absent")
        return 0
    paths = {path.name: path for path in (source.NEEDS, source.SOURCE_USE,
                                        source.CONTROLLER, source.WORLD,
                                        CAPTURE, PERCEPTION)}
    with tempfile.TemporaryDirectory(prefix="sao-transfer-lifecycle-") as temporary:
        work = pathlib.Path(temporary)
        _compile(work, source)
        checks, detail = _run(work, source, transfers)
        failed = sorted(name for name, value in checks.items() if value != "true")
        if set(checks) != EXPECTED or failed:
            print(f"FAIL provisioning lifecycle missing={sorted(EXPECTED-set(checks))} "
                  f"extra={sorted(set(checks)-EXPECTED)} failed={failed}")
            print(detail[-2500:])
            return 1
        print(f"PASS {len(checks)} production provisioning lifecycle/capture cases")
        for name, filename, before, after, expected in MUTATIONS:
            text = paths[filename].read_text(encoding="utf-8")
            if text.count(before) != 1:
                print(f"FAIL lifecycle control anchor count: {name}")
                return 1
            altered, detail = _run(work, source, transfers,
                                   {filename: text.replace(before, after, 1)})
            if set(altered) != EXPECTED or altered.get(expected) != "false":
                print(f"FAIL lifecycle mutation {name} did not flip {expected}")
                print(detail[-2500:])
                return 1
            print(f"PASS lifecycle mutation {name}: {expected}=false")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/sweep/provisioning_lifecycle_cases.lua
(function()
    if __productionPerceptionCases then
        local WS, P = SAO.WorldSources, SAO.Perception
        local checks = {}
        local function check(name, value)
            checks[#checks+1] = name .. "=" .. tostring(value and true or false)
        end
        __stores = {}
        __records = {a={id="a",forename="Ada",surname="Stone",occupation="farmer"},
            b={id="b",forename="Ben",surname="North",occupation="farmer"}}
        __bodies = {a=__newBody(8,8),b=__newBody(8,8)}
        SAO.Body.active = __bodies
        P.beliefs = {}
        __standingAllowed = true
        local building = {id=42,cx=8,cy=8,minX=8,minY=8,maxX=10,maxY=9}
        SAO.Places.at = function(x,y)
            if x == 8 and y == 8 then return building end
        end
        SAO.Standing.claimOf = function() return nil end
        SAO.Standing.groupOf = function() return nil end
        SAO.Standing.insideClaim = function() return false end
        WS.resetRuntime()
        WS.applySnapshot(WS.parse(__acquire_pre))
        local anchor = {id="source:C:food-token:0",sourceId="C:food-token:0",
            cx=8,cy=8,z=0,minX=8,minY=8,maxX=9,maxY=9}
        local learned = P.learnInspectedSource("a",anchor,"C:food-token:0",4800,
            "native-transfer-inspection")
        P.learnInspectedSource("a",anchor,"C:food-token:0",4801,"observed-transfer")
        local row = P.knownPlaces("a",true)[anchor.id]
        check("production_inspection_geometry",learned and row and row.cx == 8
            and row.cy == 8 and row.z == 0 and row.minX == 8 and row.minY == 8
            and row.maxX == 9 and row.maxY == 9 and row.sourceId == anchor.sourceId)
        local recalled = WS.nearestObserved("a",8,8,"food",10)
        check("production_anchor_recall",recalled and recalled.id == anchor.id
            and recalled.sourceId == anchor.sourceId and recalled.cx == 8 and recalled.cy == 8
            and WS.nearestObserved("b",8,8,"food",10) == nil)
        local _, _, _, scopedFacts = WS.beliefSnapshot(anchor)
        local offered = recalled and WS.actionOptions(recalled,"food","a",__bodies.a,1,"standing")
        check("production_anchor_exact_scope",row and row.sourceFacts
            and row.sourceFacts[anchor.sourceId] and not row.sourceFacts["C:neighbor:0"]
            and scopedFacts[anchor.sourceId] and not scopedFacts["C:neighbor:0"]
            and offered and #offered.options == 1
            and offered.options[1].parameters.sourceId == anchor.sourceId)
        local reservation = offered and WS.beginAction(recalled,"food","a",__bodies.a,
            1,"standing",offered.options[1])
        check("production_consume_anchor",reservation and reservation.operation == "consume"
            and reservation.placeId == anchor.id and reservation.placeSourceId == anchor.sourceId
            and reservation.sourceId == anchor.sourceId and reservation.itemId == 101)
        if reservation then WS.release(reservation.id,"fixture-complete") end
        local attachment = SAO.PlaceAttachment.of("a")
        local knownCount = 0
        for _ in pairs(P.knownPlaces("a")) do knownCount = knownCount + 1 end
        check("production_inspection_no_attachment",row and (row.visits or 0) == 0
            and knownCount == 0 and #P.returnsOf({"a"}) == 0
            and attachment and attachment.knownPlaces == 0 and attachment.visitedPlaces == 0
            and attachment.attachment == 0 and attachment.claimKind == "none")
        local unavailable = WS.nearestBelieved("a",8,8,"food",10)
        WS.applySnapshot(WS.parse(__anchor_accessible))
        P.learnInspectedSource("a",anchor,anchor.sourceId,4802,"native-transfer-inspection")
        local executable = WS.nearestBelieved("a",8,8,"food",10)
        check("production_anchor_access_route",unavailable == nil and executable
            and executable.id == anchor.id and executable.sourceId == anchor.sourceId)
        P.learnBuilding("a",building,4803,"observed-building")
        local known = P.knownPlaces("a")
        local ranking = P.returnsOf({"a"})
        attachment = SAO.PlaceAttachment.of("a")
        check("production_building_visit_preserved",known[42] and known[42].visits == 1
            and known[anchor.id] == nil and #ranking == 1 and ranking[1].id == 42
            and attachment.knownPlaces == 1 and attachment.visitedPlaces == 1)
        return table.concat(checks,"|")
    end
    local WS, SU, Ctl = SAO.WorldSources, SAO.SourceUse, SAO.Controller
    local checks, diagnostics = {}, {}
    local function check(name, value)
        checks[#checks+1] = name .. "=" .. tostring(value and true or false)
        if not value then
            local pending = WS.pendingActionFor("a")
            diagnostics[#diagnostics+1] = "CASE " .. name .. " pending=" .. tostring(pending and pending.phase)
                .. " holder=" .. tostring(__holderState) .. " moves=" .. tostring(__nativeMoves)
            local durable = __stores.SurvivorAwareness_WorldSources or {}
            for _, receipt in pairs(durable.results or {}) do
                diagnostics[#diagnostics+1] = "RESULT " .. tostring(receipt.status) .. " " .. tostring(receipt.detail)
            end
        end
    end
    local function reset(operation)
        __stores, __records, __bodies, __known, __places = {}, {}, {}, {}, {}
        SAO.Body.active, SAO.Body.foreign = __bodies, {}
        SAO.Perception.beliefs = {a={evidence={value="before"}}}
        local body = __newBody(8,8)
        __inventory, __sourceContainer, __permissionContainer, __originContainer = {}, {}, {}, {}
        body.getInventory = function() return __inventory end
        __records.a = {id="a",forename="Ada",surname="Stone",occupation="farmer",
            profile={note="before"}}
        __bodies.a = body
        Ctl.agents = {a={rec=__records.a,state="SOURCEUSE"}}
        __operation, __sourceItem = operation, {id=101,type="Base.Apple"}
        __observeText = operation == "store" and __store_pre or __acquire_pre
        local row = operation == "store" and __store_row or __acquire_row
        __offerText = row .. "\n" .. __observeText
        __carriedText = operation == "store" and row or ""
        __carriedItem = operation == "store" and __sourceItem or nil
        __destinationItem = nil
        __holderState = operation == "store" and "CARRIED" or "SOURCE"
        __bindAnswer, __targetAnswer = "BOUND:8:8:0", "READY:8:8:0:8:8:0"
        __queueReject, __busy, __queued, __queueCalls = false, false, nil, 0
        __standingAllowed, __accessible, __nativeValid = true, true, true
        __withinReach, __routeAllowed, __attemptAllowed = true, true, true
        __lastOrder, __attempt, __searchedRadius = nil, nil, nil
        __discoveredSource = "8:8:0:Apple"
        __bindCalls, __clears, __nativeMoves, __consumeCalls = 0, 0, 0, 0
        __xpCalls, __voiceCalls, __trustCalls, __debtCalls = 0, 0, 0, 0
        __groupByActor, __groupClaims, __standingStoreUnavailable = nil, nil, false
        __lastLearn = nil
        __move = function()
            __observeText = operation == "store" and __store_post or __acquire_post
            __carriedItem = operation == "acquire" and __sourceItem or nil
            __destinationItem = operation == "store" and __sourceItem or nil
            __carriedText = operation == "acquire" and __acquire_row or ""
            __holderState = "TRANSFERRED"
        end
        WS.resetRuntime()
        return body
    end
    local function begin(body, operation, context)
        local ok, why = SU.beginTransfer("a",body,"food","standing",__sourceItem,
            __sourceContainer,operation,context or {purpose="forage"})
        local reservation = WS.pendingActionFor("a")
        return ok, reservation, why
    end
    local function outcome(reservation)
        return reservation and WS.actionOutcome(reservation.id,"a") or nil
    end
    local function moveAndFinish(body)
        if __queued then __queued:transferItem(__sourceItem) end
        __busy, __queued = false, nil
        return SU.tick("a",body)
    end
    local function reload(body)
        __busy, __queued = false, nil
        SAOJavaBridge:clearWorldSourceAction(body)
        WS.resetRuntime()
        WS.reconcileReservations()
        local restored = __newBody(8,8)
        restored.getInventory = function() return __inventory end
        __bodies.a = restored
        return restored
    end

    local body = reset("acquire")
    local ok, reservation = begin(body,"acquire")
    local acquired = moveAndFinish(body)
    local receipt = outcome(reservation)
    check("acquire_performed",ok and acquired == "completed" and receipt
        and receipt.operation == "acquire" and receipt.observedQuantity == 1
        and WS.source("C:food-token:0").items["101"] == nil
        and __carriedItem == __sourceItem and __nativeMoves == 1)
    local acquireNoUse = __consumeCalls == 0 and __records.a.lastFoodDay == nil
        and __records.a.lastWaterDay == nil

    body = reset("store")
    ok, reservation = begin(body,"store",{deliveryGroup="home",requestedByGroup="home"})
    check("queued_is_not_performed",ok and __busy and outcome(reservation).status == "pending"
        and WS.source("C:food-token:0").items["101"] == nil and __xpCalls == 0)
    check("queue_has_no_social_credit",__trustCalls == 0 and __debtCalls == 0
        and __voiceCalls == 0 and __xpCalls == 0)
    local stored = moveAndFinish(body)
    receipt = outcome(reservation)
    check("store_performed",stored == "completed" and receipt
        and receipt.operation == "store" and receipt.observedQuantity == 1
        and WS.source("C:food-token:0").items["101"] ~= nil and __carriedItem == nil)
    check("no_consumption_or_need_credit",acquireNoUse and __consumeCalls == 0
        and __records.a.lastFoodDay == nil and __records.a.lastWaterDay == nil)
    local creditBefore = __xpCalls
    local repeated = Ctl.provisioningCompleted("a",body,receipt,reservation)
    check("completed_work_credit_once",creditBefore == 1 and __xpCalls == 1
        and __xpPerk == "Fitness" and __xpAmount == 1 and not repeated)
    local creditOrder = __records.a.provisioningCreditOrder
    local badActor = Ctl.provisioningCompleted("a",body,
        {actorId="other",status="completed",order=999},reservation)
    local badStatus = Ctl.provisioningCompleted("a",body,
        {actorId="a",status="pending",order=999},reservation)
    local oldOrder = Ctl.provisioningCompleted("a",body,
        {actorId="a",status="completed",order=0},reservation)
    check("credit_actor_status_order",not badActor and not badStatus and not oldOrder
        and __records.a.provisioningCreditOrder == creditOrder and __xpCalls == creditBefore)
    local physical = __known.a and __known.a[reservation.placeId]
    check("physical_anchor_refresh",reservation.placeSourceId == "C:food-token:0"
        and physical and physical.sourceFacts["C:food-token:0"].revision == "store-r2"
        and __lastLearn.provenance == "observed-transfer"
        and not physical.sourceFacts["C:neighbor:0"])

    body = reset("store")
    ok, reservation = begin(body,"store")
    __busy, __queued = false, nil
    check("queue_without_move_releases",SU.tick("a",body) == "released"
        and outcome(reservation).status == "released" and __nativeMoves == 0
        and __xpCalls == 0 and #WS.completedResults("provisioning") == 0)

    body = reset("store")
    __queueReject = true
    ok, reservation = begin(body,"store")
    check("rejected_enqueue",not ok and not __busy and not reservation
        and __queueCalls == 1 and __nativeMoves == 0 and __xpCalls == 0
        and #WS.completedResults("provisioning") == 0)

    body = reset("acquire")
    ok, reservation = begin(body,"acquire")
    SU.interrupt("a",body,"threat")
    check("interrupt_before_move",outcome(reservation).status == "released"
        and __carriedItem == nil and __nativeMoves == 0 and __xpCalls == 0)
    body = reset("store")
    ok, reservation = begin(body,"store")
    __queued:transferItem(__sourceItem)
    SU.interrupt("a",body,"threat")
    check("interrupt_after_move",outcome(reservation).status == "completed"
        and __nativeMoves == 1 and __xpCalls == 1)

    body = reset("store")
    ok, reservation = begin(body,"store")
    body = reload(body)
    local state = SU.resume("a",body)
    check("reload_before_move",state == "SOURCEUSE" and __bindCalls == 2
        and __queueCalls == 2 and __nativeMoves == 0
        and WS.pendingActionFor("a").id == reservation.id)
    body = reset("store")
    ok, reservation = begin(body,"store")
    __queued:transferItem(__sourceItem)
    body = reload(body)
    SU.resume("a",body)
    local afterReload = outcome(reservation)
    SU.resume("a",body)
    check("reload_after_move_once",afterReload and afterReload.status == "completed"
        and __queueCalls == 1 and __nativeMoves == 1 and __xpCalls == 1
        and #WS.completedResults("provisioning") == 1)

    body = reset("store")
    ok, reservation = begin(body,"store")
    body = reload(body)
    __holderState = "UNAVAILABLE"
    local waiting = SU.resume("a",body)
    check("unavailable_destination_pending",waiting == "SOURCEUSE"
        and SU.tick("a",body) == "pending" and outcome(reservation).status == "pending"
        and __queueCalls == 1 and __nativeMoves == 0)
    __holderState = "CONFLICT"
    __destinationItem = __sourceItem
    local conflictState = SU.resume("a",body)
    check("both_holders_conflict",conflictState == nil
        and outcome(reservation).status == "conflict" and __xpCalls == 0)
    body = reset("store")
    ok, reservation = begin(body,"store")
    body = reload(body)
    __carriedItem, __destinationItem = nil, nil
    conflictState = SU.resume("a",body)
    check("neither_holder_conflict",conflictState == nil
        and outcome(reservation).status == "conflict" and __xpCalls == 0)

    body = reset("store")
    ok, reservation = begin(body,"store")
    local acceptedAction = __queued
    __standingAllowed = false
    acceptedAction:transferItem(__sourceItem)
    __busy = false
    SU.tick("a",body)
    check("standing_revocation_prevents_mutation",ok and __nativeMoves == 0
        and acceptedAction.dontAdd == true and outcome(reservation).status == "released"
        and __carriedItem == __sourceItem and __xpCalls == 0)

    body = reset("acquire")
    ok, reservation = begin(body,"acquire")
    __queued:transferItem(__sourceItem)
    __carriedText = string.gsub(__carriedText,"condition=10","condition=9",1)
    __busy = false
    SU.tick("a",body)
    check("acquired_contents_mismatch",outcome(reservation).status == "conflict"
        and __records.a.lastFoodDay == nil and __xpCalls == 0)

    body = reset("acquire")
    __standingAllowed = false
    ok = SAO.Needs.queueTake("a",body,{category="food",purpose="forage",admission="standing"})
    check("standing_food_refusal",not ok and not WS.pendingActionFor("a")
        and __queueCalls == 0 and __nativeMoves == 0 and __standingAdmission == "standing")
    body = reset("acquire")
    __standingAllowed = false
    ok = SAO.Needs.queueTake("a",body,{category="food",purpose="forage",admission="desperate"})
    reservation = WS.pendingActionFor("a")
    local desperateOutcome = moveAndFinish(body)
    check("desperate_food_admission",ok and reservation and reservation.admission == "desperate"
        and __standingAdmission == "desperate" and desperateOutcome == "completed"
        and __nativeMoves == 1 and __records.a.lastFoodDay == nil)
    body = reset("acquire")
    __sourceItem = {id=404,type="Base.Pills"}
    __offerText = ""
    ok = SAO.Needs.queueTake("a",body,{category="legacy"})
    check("legacy_medication_route",ok and __queueCalls == 1 and __queued.item == __sourceItem
        and not WS.pendingActionFor("a") and __consumeCalls == 0)
    body = reset("acquire")
    __offerText = ""
    ok = SAO.Needs.queueTake("a",body)
    check("legacy_missing_context",ok and __queueCalls == 1 and not WS.pendingActionFor("a"))

    body = reset("acquire")
    __withinReach, __discoveredSource = false, "12:11:2:Apple"
    local nearbyState, nearbyContext = SAO.Needs.collectNearby("a",body,20,99)
    check("nearby_approach_exact",nearbyState == "FORAGE" and nearbyContext
        and nearbyContext.category == "food" and nearbyContext.purpose == "forage"
        and nearbyContext.admission == "standing" and nearbyContext.haulRemaining == 3
        and nearbyContext.haulRadius == 4 and __searchedRadius == 20
        and __lastOrder and __lastOrder.id == "a" and __lastOrder.x == 12
        and __lastOrder.y == 11 and __lastOrder.z == 2 and __attempt
        and __attempt.x == 12 and __attempt.y == 11 and __attempt.admission == "standing"
        and __queueCalls == 0 and not WS.pendingActionFor("a") and __nativeMoves == 0)
    body = reset("acquire")
    __withinReach, __routeAllowed = false, false
    nearbyState = SAO.Needs.collectNearby("a",body,4,2)
    check("nearby_route_refusal",nearbyState == nil and __lastOrder
        and __queueCalls == 0 and not WS.pendingActionFor("a") and __nativeMoves == 0)
    body = reset("acquire")
    __withinReach, __attemptAllowed = false, false
    nearbyState = SAO.Needs.collectNearby("a",body,4,2)
    check("nearby_standing_refusal",nearbyState == nil and __attempt
        and __attempt.admission == "standing" and not __lastOrder
        and __queueCalls == 0 and not WS.pendingActionFor("a") and __nativeMoves == 0)
    body = reset("acquire")
    nearbyState, nearbyContext = SAO.Needs.collectNearby("a",body,0,0)
    reservation = WS.pendingActionFor("a")
    check("nearby_reached_transfer",nearbyState == "TAKE" and nearbyContext
        and nearbyContext.haulRemaining == 0 and nearbyContext.haulRadius == 1
        and reservation and reservation.operation == "acquire" and reservation.itemId == 101
        and __queued and __queued.item == __sourceItem and __queueCalls == 1
        and not __lastOrder and __nativeMoves == 0)

    body = reset("store")
    local choose = SU.chooseOption
    SU.chooseOption = function(offered)
        __records.a.profile.note = "during"
        SAO.Perception.beliefs.a.evidence.value = "during"
        return choose(offered)
    end
    local capture = SAODecisionCapture.beginSourceUse({runId="run-transfer",county="Knox"})
    ok, reservation = begin(body,"store")
    moveAndFinish(body)
    __records.a.profile.note = "after"
    SAO.Perception.beliefs.a.evidence.value = "after"
    local text = capture.finish()
    SU.chooseOption = choose
    local function has(needle) return string.find(text,needle,1,true) ~= nil end
    check("capture_transfer_bound",ok and has('"eventCount":1')
        and has('"captureFailureCount":0') and has('"reservationId":"' .. reservation.id .. '"')
        and has('"status":"completed"') and not has('"action-not-bound"'))
    check("capture_decision_frozen",has('"note":"before"') and has('"value":"before"')
        and not has('"note":"during"') and not has('"note":"after"'))
    check("capture_item_units",has('"quantityUnit":"item"')
        and has('"measurement":"native-item-transfer"') and has('"observedQuantity":1'))

    body = reset("acquire")
    capture = SAODecisionCapture.beginSourceUse({runId="run-bad-result",county="Knox"})
    ok, reservation = begin(body,"acquire")
    local read = WS.actionOutcome
    WS.actionOutcome = function() return nil,"store-unavailable" end
    text = capture.finish()
    WS.actionOutcome = read
    check("capture_required_result_failure",ok and has('"eventCount":0')
        and has('"captureFailureCount":1') and has('"stage":"result"')
        and not has('"status":"censored"'))

    body = reset("store")
    local age = SAO.History.ageOf
    SAO.History.ageOf = function() error("fixture age reader unavailable") end
    capture = SAODecisionCapture.beginSourceUse({runId="run-bad-person",county="Knox"})
    ok, reservation = begin(body,"store")
    text = capture.finish()
    SAO.History.ageOf = age
    check("capture_required_person_failure",ok and __busy and reservation
        and has('"eventCount":0') and has('"captureFailureCount":1')
        and has('History.ageOf'))
    return table.concat(checks,"|") .. "|DETAIL " .. table.concat(diagnostics," ; ")
end)()
